### ISSUE TITLE
Code inefficacy

### ISSUE THREAD

**Issue by 0mlml** [Login1990/CafeteriaBot#1]
I noticed that you scrape for the information about menus. You mentioned that you have had to rewrite your code due to layout changes. This could be solved easily if you requested the information from the API directly. I found an api [for the day](https://www.compass-group.fi/menuapi/day-menus?costCenter=3424&date=2023-10-13T07%3A30%3A29.247Z&language=fi) and [for the week:](https://www.compass-group.fi/menuapi/week-menus?costCenter=3424&date=2023-10-09T07%3A30%3A29.247Z&language=fi). It appears you can control the date and language through parameters also. 

Scraping is generally ill-advised due to the increased load to the destination server as you force the server to load excess information that you end up discarding.

I can open a PR later with these changes.

**Comment by Login1990** [comment]
@0mlml never knew there was API, how did you found it?

**Comment by Login1990** [comment]
If I will have time tomorrow, I will come up with implementation myself, but if you want to - I can leave it to you

**Comment by 0mlml** [comment]
There is no obviously no official API, but the data must come from somewhere. As you discovered, the initial page load does not contain the data (see [SSR](https://vuejs.org/guide/scaling-up/ssr.html)), so the data must be loaded afterwards. This is generally done with an XHR request in the site's JavaScript.

I found the endpoint by opening up the network tab in devtools and filtering by XHR to find where the information about menu items is loaded. The day-menus endpoint stuck out immediately, and the week-menus endpoint appears when clicking "Koko Viikko" (Whole week). I have attached a screenshot of the request log as a visual. 
![image](https://github.com/Login1990/CafeteriaBot/assets/63015852/33b997a5-9719-409f-bc6e-80e7a6fff9ce)

Clicking on these requests, I can see the data that is loaded onto the site, as HTML. This is not ideal, as it will require parsing the HTML for the data we want, as opposed to JSON, where the data is readily accessible.
![image](https://github.com/Login1990/CafeteriaBot/assets/63015852/661fca6c-359f-4c97-b5b2-6fe051bba83f)

Another issue I saw with your bot is how it is run. I don't know the environment that you are running it in, but the scheduler implementation is very hacky. If you are running on linux, you are in luck with the support for scheduled jobs. Some great options are [cron](https://www.geeksforgeeks.org/crontab-in-linux-with-examples/) and [Systemd](https://wiki.archlinux.org/title/systemd/Timers). On Windows, you are stuck with task scheduler, but it is a fairly intuitive GUI.

Your implementation of the scheduled scraper has the right idea though, caching the data to disk to avoid spamming the server. I think that rather than writing to disk, which is fairly clunky, you could save the data in memory along with a timestamp of when the data was last fetched. The general idea would be as follows:
```py
latest_menu_data = None
last_menu_fetch = 0
menu_fetch_timeout = 3600  # 1 hour

def fetch_menu_data():
    global latest_menu_data, last_menu_fetch
    # Your code to fetch and parse the menu data
    last_menu_fetch = time.time()

def get_menu_data():
    if time.time() - last_menu_fetch > menu_fetch_timeout:
        fetch_menu_data()
    return latest_menu_data

# When accessing the data:
menu_data = get_menu_data()
```

It would simplify your code drastically by not having to deal with encoding and decoding JSON, writing and reading to disk, and wouldn't really use that much more bandwidth or memory.

**Comment by Login1990** [comment]
@0mlml 
Thanks for the input. I agree that scraping is at most dubious, and I am currently implementing the API requests.
I also think that scheduling through python script is not optimal, I will try to get around cron.
However the menu is updated only once a week, on friday, so updating it every hour (if there is a user however), takes a bit more than doing it once a week or a day

I think I will stick with storing a .json for a time being

**Comment by Login1990** [comment]
Code updated
